Mathlib/Algebra/Module/Character.lean

```diff
@@ -0,0 +1,140 @@
+import Mathlib.Algebra.Category.GroupCat.Injective
+import Mathlib.Topology.Instances.AddCircle
+import Mathlib.Topology.Instances.Rat
+
+open CategoryTheory
+
+universe u v w w'
+
+variable (R : Type u) [CommRing R]
+variable (M : Type w) [AddCommGroup M] [Module R M]
+variable (N : Type w) [AddCommGroup N] [Module R N]
+
+
+def CharacterModule : Type w :=
+M →ₗ[ℤ] (AddCircle (1 : ℚ))
+
+instance : FunLike (CharacterModule M) M (fun _ => AddCircle (1 : ℚ)) where
+  coe (f : M →ₗ[ℤ] (AddCircle (1 : ℚ))) m := f m
+  coe_injective' _ _ h := LinearMap.ext fun _ => congr_fun h _
+
+instance : AddMonoidHomClass (CharacterModule M) M (AddCircle (1 : ℚ)) where
+  coe f := f
+  coe_injective' _ _ h := FunLike.ext _ _ fun _ => congr_fun h _
+  map_add f := f.map_add
+  map_zero f := f.map_zero
+
+instance : AddCommGroup (CharacterModule M) := by
+  delta CharacterModule
+  infer_instance
+
+instance : Module R (CharacterModule M) where
+  smul r l :=
+  { toFun := fun x => l (r • x)
```

Comment:
I'm puzzled by how the current state of this PR (and #8906) departed from this original commit. In this commit, the domain M is a (R,Z)-bimodule and the codomain `AddCircle (1 : ℚ)` is only a Z-module, and the R-action on `CharacterModule M` is on the domain (of course). As the PR currently stands, both the domain and the codomain are R-modules and there isn't a second ring, and you're using LinearMap.module to define the R-module structure on `CharacterModule R M D`, which acts on the codomain (but it's equivalent to acting on the domain, because you consider R-linear maps not Z-linear maps).

Are you sure the current definition can still help you prove things about flatness? Do you have it in some branch? I'd be curious to see. My suspicion is that Matlis duality doesn't fit in the same framework as what you want to do.

Reply:
See #9120.

Reply:
We can still force the proof to work because `M →ₗ[ℤ] ℚ ⧸ ℤ` is still an `R`-module, but `CharcterModule.(un)curry, cong, homEquiv` cannot be used directly because they will only be `ℤ`-linear, but since they are still true, we can just make another copy of these constructions and use `CharacterModule.unitRationalCircle.(un)curry, cong, homEquiv` which are `R`-linear to prove injectiveness.